modules/dists/BlockDist.chpl

```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
   return true;
 }
 
+proc BlockArr.canDoOptimizedSwap(other) {
+  var domsMatch = true;
+  var domShapesMatch = true;
+
+  if this.dom != other.dom { // no need to check if this is true
+    if domsMatch {
+      for param i in 0..this.dom.rank-1 {
+        if this.dom.whole.dim(i) != other.dom.whole.dim(i) {
+          if this.dom.whole.dim(i).size != other.dom.whole.dim(i).size {
+            domShapesMatch = false;
+          }
+          domsMatch = false;
+        }
+      }
+    }
+  }
+
+  if domsMatch {
+    // distributions must be equal, too
+    return this.dom.dist.dsiEqualDMaps(other.dom.dist);
+  }
+  // we can handle different domains if:
+  // their shapes match, and
+  // they also match their respective distributions' boundingBoxes
+  else if domShapesMatch {
+    // TODO: this rule can be relaxed a bit.  ie. arrays on the following
+    // domains have matching shaped and can be swapped with a pointer swap, but
+    // that's not what happens because of this line is only checking for a very
+    // specific case.
+    //
+    // var d1 = {0..9 by 2 align 0} dmapped Block({1..10});
+    // var d2 = {0..9 by 2 align 1} dmapped Block({1..10});
+    return this.dom.dist.boundingBox == this.dom.whole &&
+           other.dom.dist.boundingBox == other.dom.whole;
+  }
+  return false;
+}
+
+// <=> implementation uses Reflection to check whether this resolves, so
+// it is OK for this to not resolve without debugOptimizedSwap.
+proc BlockArr.doiOptimizedSwap(other) where debugOptimizedSwap {
+  writeln("BlockArr doing unoptimized swap. Type mismatch");
+  return false;
+}
+
 // Block1 <=> Block2 
-proc BlockArr.doiOptimizedSwap(other) {
-  if(this.dom.dist.dsiEqualDMaps(other.dom.dist)) {
+proc BlockArr.doiOptimizedSwap(other: this.type) 
+    where this.stridable == other.stridable &&
+          this.rank == other.rank {
```

Comment:
Is this where clause redudant with the `this.type` constraint?  My intuition says "yes", but since it's here, I worry that I'm missing something.  If so, I think it should be removed, both for simplicity, to avoid spreading superstitious coding if people start worrying that the type constraint isn't sufficient, and because of the aforementioned cost of evaluating where clauses in the compiler.

So then the "resolution performance" question for me becomes "is choosing between two overloads cheaper or more expensive than having a single generic function that we instantiate for each type we try to resolve it with?" (where again, I'm unsure... but I mention it now that I see it was generic before, and because I think the logic / choices would be easier to follow with a single routine containing control flow vs. multiple overloads that rely upon formal type constraints and where clauses to understand.  i.e., if compilation time was not an issue or we knew that it was no better, I'd definitely put everything back into a single routine again for code clarity).  I'm not at all insistent on this, just commenting to see what you think.

Reply:
The where clause should go away, I believe. 

But your question is interesting. My intuition as I mentioned above is that multiple overloads is better as you wouldn't instantiate a big function only to fold some param ifs later on. That being said, I don't trust in my knowledge on the order of operations during function resolution. Though it makes sense that first you need to instantiate the function than fold ifs in instantiations?

Reply:
There is a bug (quirk?) with canResolveMethod that these where clauses should remain or I should change `this.type` in the formal declaration to something more explicit.

```chapel
use Reflection;

record R {
  var x;

  proc foo(r: this.type) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }

  proc bar(r: R(int(64))) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }

  proc baz(r: R(this.x.type)) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }
}

var r1 = new R(5);
var r2 = new R(5.2);

compilerWarning("Can resolve foo? ", canResolveMethod(r1, "foo", r2):string);
compilerWarning("Can resolve bar? ", canResolveMethod(r1, "bar", r2):string);
compilerWarning("Can resolve baz? ", canResolveMethod(r1, "baz", r2):string);
```

Generates:

```
$CHPL_HOME/canResolveBug.chpl:6: In method 'foo':
$CHPL_HOME/canResolveBug.chpl:7: warning: Resolving foo. this.type: R(int(64)) r.type: R(real(64))
  $CHPL_HOME/modules/standard/Reflection.chpl:290: called as R(int(64)).foo(r: R(real(64))) from function 'canResolveMethod'
  $CHPL_HOME/canResolveBug.chpl:1: called as canResolveMethod(obj: _any, param fname = "foo": string, args(0): _any)
note: generic instantiations are underlined in the above callstack
$CHPL_HOME/canResolveBug.chpl:1: In module 'canResolveBug':
$CHPL_HOME/canResolveBug.chpl:25: note: Can resolve foo? true
$CHPL_HOME/canResolveBug.chpl:26: warning: Can resolve bar? false
$CHPL_HOME/canResolveBug.chpl:27: warning: Can resolve baz? false
```

Where if the formal type is `this.type` we don't match against the generic fields and try to resolve the function. This causes resolution errors in this branch. I'm recording this here to investigate further soon.

Reply:
OTOH, this made me realize that we should probably prevent the optimized path from executing if the arrays element types are different. As it can change the type mismatch error messages in a bad way. Last commits make that change

Reply:
I recorded the reflection problem in https://github.com/chapel-lang/chapel/issues/16673

Reply:
OK, thanks!